### DISCUSSION TITLE
RPi PicoW and ILI9341+XPT2046

### DISCUSSION THREAD

**Discussion by washburnit** [Bodmer/TFT_eSPI#2230] [Q&A - Touch]
Hi,
I bought the display in the subject, I'm using it with a RPi PicoW.
That display is equipped with a touchscreen controller based on a XPT2046 and a SD reader.
I was able to setup the ILI9341 correctly using the "Setup61_RP2040_ILI9341_PIO_SPI.h", I connected the pins as follows:

#define TFT_MISO  5 // MISO is not used or supported
#define TFT_MOSI  2
#define TFT_SCLK  3
#define TFT_CS     0  // Chip select control pin
#define TFT_DC     1  // Data Command control pin
#define TFT_RST   4  // Reset pin (could connect to Arduino RESET pin)

#define TFT_BL   6           // Optional LED back-light control pin
#define TFT_BACKLIGHT_ON HIGH  // Level to turn ON back-light (HIGH or LOW)

I control the GPIO6 (backlight) through software.
All the "graphics" is working very good but I would like to use the touch feature.
I noticed that there is a "Setup60_RP2040_ILI9341.h" with 

#define TOUCH_CS 16     // Chip select pin (T_CS) of touch screen

that I connected to GPIO16.
I modified the "User_Setup_Select.h" including the file above but when I compile the modified sketch the display appears filled with white and nothing happens, if I revert to the "Setup61..." everything works.
What am I doing wrong?
The other "touch pins" labelled on the PCB as T_IRQ, T_DO, T_DIN, T_CLK where must be connected?
Thanks you, regards.

Roberto

**Comment by Bodmer** [comment]
A touch controller is not supported when a PIO based SPI interface is used.

The Touch controller is a SPI device and therefore must be wired to the processor SPI pins and a chip select pin allocated.

Example connections:
```
T_DO  -> MISO pin (e.g. TFT_MISO)
T_DIN -> MOSI pin (e.g. TFT_MOSI)
T_CS  -> pin of you choice specified in setup file as TOUCH_CS
T_CLK -> SCLK pin (e,g, TFT_SCLK)
T_IRQ -> not connected
```

**Reply by dslocum** [reply]
I have a Seeed XIAO RP2040 with the above connections.  The display works fine with any of the demos I've tried, but I can't get any touch examples to run.

I'm currently using Test_Touch_Controller and the Serial monitor shows all 3 parameters = "0".  On a o'scope there are reasonable signals on the pins and I even get outputs on the MISO line that appear to have data changing as I move my finger / stylus.
![Screenshot 2023-08-19 131823](https://github.com/Bodmer/TFT_eSPI/assets/16817473/db4359b5-e3cf-4af2-87dd-091bb9f01f69)

Anything else I could look for?

**Comment by washburnit** [comment]
Thanks for the prompt reply.
As for the configuration file, should I change the configuration file from the "Setup61...." to "Setup60..." then?
With this last I'm experiencing an issue (white display).

Roberto

**Reply by Bodmer** [reply]
Use setup60 with a new pin allocated to the touch controller chip select. To keep things simple use the ORIGINAL setup pin numbers. There are constraints on what pin combination will work, see pin capability combinations here:
https://www.raspberrypi.com/documentation/microcontrollers/raspberry-pi-pico.html#pinout-and-design-files

Note also if you use a port 1 SPI pin set you must also define the port number in the setup file. Hence it is best to use the original tested setup files until these complexities are understood.

**Comment by washburnit** [comment]
Thanks, it worked...but...some small "artifacts" appeared on the screen like black lines on coloured area or font chars not drawn correctly or incomplete.
I tried to lower the SPI_FREQUENCY to  62500000 instead than the original SPI_FREQUENCY  70000000 but some artifacts keep appearing, what should I check?

Roberto

**Reply by Bodmer** [reply]
It sounds like the SPI clock speed is too high. Bear in mind that the maximum recommended clock frequency for the ILI9341 is SPI mode is 10MHz (100ns clock period). Typically transfer relability degrades above about 50MHz but his varies between displays. Some of these displays seem to be Chinese clones of the ILI9341.

The SPI clock frequencies are integer divisions of the CPU clock frequency and there is a complication that the clock source get changed to a much lower value if a 133MHz CPU is selected, so you do not always get what is requested.

**Comment by washburnit** [comment]
I've set it at 20MHz, the rendering speed is worst but still acceptable...I will try to increase the frequency gradually until I found the reliable speed.

**Reply by Bodmer** [reply]
I have not seen any problems at 40MHz on my ILI9341 displays, when short wires are used. YMMV.
